### PR TITLE
Remove the automatic creation and loading of the default wallet

### PR DESCRIPTION
**Pull request by achow101**
Instead of automatically creating and loading a default wallet, users should instead explicitly create their wallet or load it on start.

Builds on #19754 which provides the `load_on_startup` behavior for the GUI.

### CODE REVIEW

File: src/wallet/init.cpp
Comment (promag) [inline src/wallet/init.cpp]: Unless I'm missing something, this is breaking change. Why not change the behavior from "load or create each -wallet" to "just load each -wallet"? Then to create a wallet there is the option in the GUI and `createwallet` in RPC.

**Comment by Sjors** [comment]
Partial concept ACK on:
* not having a default wallet with `bitcoind`
* storing opened wallets in `bitcoin_rw.conf`

Concept on the fence on make this a breaking change for `bitcoind`. I.e. maybe we should still load an existing `<datadir>/wallet.dat` and `<datadir>/wallet.dat`, but not create one.

Concept NACK on:
* having QT start with a blank wallet, where the user is instructed to go the file menu and choose Create or Open Wallet
* not loading the default wallet in QT after upgrading: that's a guarantee for user heart attacks

I think the best solution is here is to pull the create wallet UI into the startup wizard, as an additional section. The user can then just click next and it creates the default wallet. Since it takes at least a few seconds, this gives the RNG a bit more bit time too (I assume the IBD network traffic should generate some extra entropy).

An easier but still acceptable solution is a big Create Wallet button in the Overview tab (when no wallet is loaded), as well as point out that they can load an existing wallet (less likely for first time users).

Some issues:
* opening and closing wallets in QT updates `bitcoin_rw.conf`, which is good but:
   * why use a new field `loadedwallet`? adding and removing `wallet=` entries seems cleaner.  It won't work for wallets already in bitcoin.conf, but that's fine (we can warn the user in that case).
   * I suggest doing this as soon as the open / close succeeds, rather than at exit

**Comment by achow101** [comment]
> An easier but still acceptable solution is a big Create Wallet button in the Overview tab (when no wallet is loaded), as well as point out that they can load an existing wallet (less likely for first time users).

I've done this.

> not loading the default wallet in QT after upgrading: that's a guarantee for user heart attacks

I've changed that back to the original behavior except that it only loads the default wallet if it is available. It will not create it.

>     * why use a new field `loadedwallet`? adding and removing `wallet=` entries seems cleaner.  It won't work for wallets already in bitcoin.conf, but that's fine (we can warn the user in that case).

#11082 doesn't let me set the multiple of the same argument.

**Comment by Sjors** [comment]
Much better, thanks. Both bitcoind and QT now load the default wallet if it exists, and the QT create button is more clear.

Can you add screenshots of the UX to the PR description?

The initial download overlay actually hides the wallet functionality, which has the nice side-effect of collecting some entropy. Although I'd rather work on making the initial experience such that users can get started more quickly, that's a whole different story. 

schermafbeelding 2019-02-22 om 09 38 29

When the user hides the sync they see this:
schermafbeelding 2019-02-22 om 09 38 55

A bit ugly, but I think the button makes it super clear what to do. Nit: reverse the order, i.e. suggest creating a wallet and then point out that the user can also load an existing wallet.

Regarding the wallet creation dialog, that can be discussed in #15450, but it's important that we don't overwhelm a first time user with advanced options. One thing though: what to do with the default name, at the risk of bike shedding...
schermafbeelding 2019-02-22 om 09 40 37

I suggest we hide the wallet name field for the first wallet, or make it really clear that it's totally optional and that you can't (easily) change it (yet).


As for rw_config, which has been in limbo for quite a long time, let's start with reviewing #14866 by @AkioNak which makes the evaluation of `bitcoin.conf` (includes) a bit more sane. We should expand the functional test suite to more thoroughly cover the expected behavior. That in turn should make it easier to merge rw_config.

Found a small bug:
```
src/bitcoin-cli -testnet listwallets
[
]

src/bitcoin-cli -testnet createwallet ""
error code: -4
error message:
Wallet  already exists
```

**Comment by jnewbery** [comment]
concept ACK

File: src/wallet/init.cpp
Comment (practicalswift) [inline src/wallet/init.cpp]: Nit: Redundant initialisation :-)

File: src/wallet/load.cpp
Comment (luke-jr) [inline src/wallet/load.cpp]: What if the wallet name has a comma? Would prefer to see rwconf extended to multiple `wallet=` fields...

File: src/wallet/load.cpp
Comment (luke-jr) [inline src/wallet/load.cpp]: Should just use `wallet=`?

File: src/wallet/load.cpp
Comment (achow101) [reply src/wallet/load.cpp]: So would I. I would prefer that #11082 implements it that way instead of requiring me to change it as I am not familiar with that code.

File: src/wallet/load.cpp
Comment (achow101) [reply src/wallet/load.cpp]: That will cause some problems as is since it would be one string that represents multiple wallets. If/when rwconf allows repeated fields, then it can be `wallet=`.

**Comment by achow101** [comment]
Rebased this on top of #15937

File: src/wallet/init.cpp
Comment (ryanofsky) [inline src/wallet/init.cpp:121]: In commit "Do not create default wallet" (6feed9508a225276cc2264b7d19b151a62deecea)

Was experimenting with this PR and would suggest the following fixup: 973bb7e652c4ce10ee536e9967a88c80f6dbe2e0 ([branch](https://github.com/ryanofsky/bitcoin/commits/review.15454.2-edit)). SoftSet doesn't really work here for a persistent setting. I'll backport this fixup to #15937 so the diff here will be smaller, and the only code change this should have to make is adding ` && WalletLocation("").Exists()`

**Review by ryanofsky — APPROVED** [review]
Approach ACK 830351b6531eb532830cc6f6523232f16b8958d1. Simple & good change overall

File: src/wallet/init.cpp
Comment (ryanofsky) [reply src/wallet/init.cpp:121]: re: https://github.com/bitcoin/bitcoin/pull/15454#discussion_r445559622

> Was experimenting with this PR and would suggest the following fixup: [973bb7e](https://github.com/bitcoin/bitcoin/commit/973bb7e652c4ce10ee536e9967a88c80f6dbe2e0) ([tag](https://github.com/ryanofsky/bitcoin/commits/review.15454.2-edit.1))

Feel free to steal updated version of this PR with fix above squashed and newer changes from 15937:
https://github.com/ryanofsky/bitcoin/commits/review.15454.2-edit

**Comment by achow101** [comment]
Updated with the changes from #15937

**Comment by Sjors** [comment]
Could use a fresh rebase in light of discussion in bitcoin-core/gui#26

**Comment by achow101** [comment]
> Could use a fresh rebase in light of discussion in [bitcoin-core/gui#26](https://github.com/bitcoin-core/gui/issues/26)

Waiting on #15937 rebase

**Review by ryanofsky — APPROVED** [review]
Code review ACK efd93dec07906b4f28e0a9b5620c2c426c9e80d2 last 3 commits. No changes since last review other than rebase

**Review by ryanofsky — APPROVED** [review]
Code review ACK 546bdce4ece60635a1eb2fa9d46d82ae41aeee24 last 3 commits. No changes since last review other than rebase

File: src/qt/walletframe.cpp
Comment (flack) [inline src/qt/walletframe.cpp:37]: wouldn't it be nicer to just make a `open_wallet_button` here like it is done below for the create button? That would save the user a bit of mousing around. Also, it would remove the headache of making sure that `File > Open Wallet` stays consistent with how the menu items are actually called

Comment (achow101) [reply src/qt/walletframe.cpp:37]: It would, but the way that opening wallets works now would require more extensive changes with code that I'm not too familiar with.

Comment (flack) [reply src/qt/walletframe.cpp:37]: if this were Javascript, I would simply make a button that clicks the menu item. But I guess that won't fly in Qt :-)

Comment (achow101) [reply src/qt/walletframe.cpp:37]: The menu item doesn't open a selection dialog, it opens a submenu listing all of the wallets. That doesn't really work here.

Comment (promag) [reply src/qt/walletframe.cpp:37]: Do you think we should replace the submenu with a dialog?

Comment (achow101) [reply src/qt/walletframe.cpp:37]: I suppose this could just be a dropdown listing all of the available wallets. But, as I said, I leave this up to someone more familiar with the GUI to do.

Comment (flack) [reply src/qt/walletframe.cpp:37]: if it were up to me I'd say a simple filepicker would totally suffice. It's self-explanatory and flexible, and the more fancy stuff is still available from the menu for those who want it

Comment (achow101) [reply src/qt/walletframe.cpp:37]: We would need #15204 to do a file picker type thing.

Comment (flack) [reply src/qt/walletframe.cpp:37]: ah, sorry, I somehow thought that was already merged

Comment (promag) [reply src/qt/walletframe.cpp:37]: @flack see https://github.com/bitcoin/bitcoin/pull/15204#issuecomment-539932726.

**Comment by achow101** [comment]
I've changed the way that the GUI handles the `load_on_start` stuff and split it into #19754

**Review by ryanofsky — APPROVED** [review]
Code review ACK a386fcd605509767b20a90719cf966e093bfa5dd. Only change since last review is rebasing and replacing bbda06f9cd1345aa80301bfc98b1cae4ec231f6d with e8bfa09b1c1888c33acfd5dcebbe136bf0d6fe82 from #19754.

re: https://github.com/bitcoin/bitcoin/pull/15454#issue-254819021

> Instead of automatically creating and loading a default wallet, users should instead explicitly create their wallet or load it on start.
> 
> This makes use of the `load_on_start` feature provided by #15937. The GUI is also changed to always do load on start for all loaded and created wallets.

It would probably be good to update PR description to mention #19754 instead of #15937

**Comment by Sjors** [comment]
Having https://github.com/bitcoin-core/gui/issues/77 implemented would go a long way towards making the "Create Wallet" dialog - which all new users have to deal with after this PR - less intimidating.

**Comment by ryanofsky** [comment]
#19754 isn't really mentioned anywhere here but the first commit in this PR is from #19754 and can be reviewed there

**Review by ryanofsky — APPROVED** [review]
Code review ACK 952bf2c9eef9d6b2f1a01142191c5f41403ed3de. No changes since last review except in first commit which is its own PR #19754

**Comment by promag** [comment]
I guess we can leave https://github.com/bitcoin/bitcoin/pull/19754#issuecomment-685858578 for another PR. But I wonder what the best approach would be. Currently we have:
```cpp
auto wallet_client = interfaces::MakeWalletClient(*node.chain, args, args.GetArgs("-wallet"));
```
Which then calls
```cpp
std::shared_ptr<CWallet> pwallet = CWallet::CreateWalletFromFile(chain, WalletLocation(walletFile), error, warnings);
```
So, for this code path, we should add some sort of flag to "create if exists"/"just load".

Alternatively we could filter settings `-wallet`, but I think that's the best approach.

**Comment by ryanofsky** [comment]
> I guess we can leave [#19754 (comment)](https://github.com/bitcoin/bitcoin/pull/19754#issuecomment-685858578) for another PR. But I wonder what the best approach would be. Currently we have:

Ideally, I don't think it should be handled in the wallet/init.cpp Construct method. That method was only meant to construct WalletClient object early on startup, not to access the filesystem or do anything slow or blocking. Also with #10102, that function runs in the node process, not the wallet process.

More ideally, handling not existing wallets would be done in wallet/load.cpp VerifyWallets function. With #19619 all that would be required is setting the `DatabaseOptions::require_existing` option and handling `DatabaseStatus::FAILED_NOT_FOUND` return code. It would be easy to implement permanently removing non-existing wallets from settings.json there, or just temporarily ignoring them.

More difficult but nicer, I think. would be to have non-existing wallets not removed from settings.json, and shown in the GUI in a disabled state, greedy out with "Wallet path was not accessible" message. That way user could be informed that wallet wasn't accessible and decide whether or not they want to keep trying to load it in the future.

**Comment by promag** [comment]
> and shown in the GUI in a disabled state

IMO more appropriate is to show a modal dialog on start (for each non existing wallet) asking if it should be removed/forgotten from settings.

**Comment by ryanofsky** [comment]
Matter of taste, I guess. I don't love modal dialogs, and if I have a multiple wallets and some of them are on an encrypted mount or external drive which is unavailable, I would rather just see the wallets listed normally, disabled and unavailable, not have to manually click through a bunch of pop-up dialogs  confirming serially that I still want to load each wallet in the future.

**Comment by promag** [comment]
How would you drop disabled wallets (from settings)?

**Comment by ryanofsky** [comment]
> How would you drop disabled wallets (from settings)?

The same way you drop an enabled wallet from settings. Switch to wallet in drop-down, go to File -> Close Wallet...

**Comment by promag** [comment]
Oh, so it would be possible to switch in the combo (not actually disabled) but then WalletFrame/WalletView should say that wallet is not available or something like that and probably disable send/receive/transactions views.

**Comment by ryanofsky** [comment]
Yes, exactly. I guess the term "disabled" is overloaded. But that's just a suggestion. Other approaches (just silently not loading non-existing wallets, silently not loading and also removing from settings, or prompting the user) all seem reasonable and probably easier to implement.

**Comment by meshcollider** [comment]
Error building after #19619 was merged:
```
wallet/init.cpp: In member function ‘virtual void WalletInit::Construct(NodeContext&) const’:
wallet/init.cpp:112:37: error: ‘WalletLocation’ was not declared in this scope
     if (!args.IsArgSet("wallet") && WalletLocation("").Exists()) {
                                     ^~~~~~~~~~~~~~
wallet/init.cpp:112:37: note: suggested alternative: ‘WalletBatch’
     if (!args.IsArgSet("wallet") && WalletLocation("").Exists()) {
                                     ^~~~~~~~~~~~~~
                                     WalletBatch
Makefile:11482: recipe for target 'wallet/libbitcoin_server_a-init.o' failed
```

**Comment by achow101** [comment]
Rebased

**Review by meshcollider — COMMENTED** [review]
Lightly tested ACK c24bb1a09cc608b499fefba2402a497cf504217d

**Review by ryanofsky — APPROVED** [review]
Code review ACK c24bb1a09cc608b499fefba2402a497cf504217d, but I think it would be good to fix up the first commit to not do file I/O in WalletInit::Construct and not hardcode "wallet.dat" berkeley wallet name outside of `bdb.cpp`, and to call existing `MakeWalletDatabase` function instead of adding a new `HasDefaultWallet` function. Suggested changes that I think would be to squash are: 421a5502ec71dff3e2d26c7f63cbd3d8da10a706 ([branch](https://github.com/ryanofsky/bitcoin/commits/review.15454.8-edit)) +32/-33 lines.

Also would also be good if the change had release notes. Off the top of my head something like "Bitcoin Core will no longer create an unnamed `""` wallet by default when no `-wallet=<name>` command line or `wallet=<name>` configuration options are used, but for backwards compatibility if a `""` wallet already exists and would have been loaded previously, it will still be loaded. Instead of automatically creating an unnamed unencrypted wallet, the GUI now prompts to create a named, encrypted wallet. Instead of only loading the unnamed `""` wallet on startup the GUI loads previously loaded wallets."

**Comment by achow101** [comment]
Squashed in @ryanofsky's commit. Also added release notes.

**Review by ryanofsky — APPROVED** [review]
Code review ACK d26f0648f1c0d1115dcb8d76e57195032b88f400. Just suggested changes to first commit (reusing MakeWalletDatabase and adding release notes), no changes to second commit

**Comment by ryanofsky** [comment]
I would really like to see this merged, because I think it's a significant interface change (or at least a notable one)  that a lot of other changes can build on:

- GUI improvements like https://github.com/bitcoin-core/gui/issues/77
- Fixing bad behavior of `settings.json` wallet list recreating removed wallets https://github.com/bitcoin/bitcoin/pull/19754#issuecomment-685858578
- Maybe turning off non-explicit wallet create entirely, and making wallet creation require a `createwallet` call or gui interaction. jb55 reported problems caused by this yesterday in IRC: http://www.erisian.com.au/bitcoin-core-dev/log-2020-09-14.html#l-355
- Test improvements: getting rid of -wallet/createwallet RPC override hacks and `""` wallet reliance in in test framework, and switching to explicit named creates. 5d2e85c5d33c3b65c79673cad5d7bc0e5875b107 starts down this road

**Comment by jonatack** [comment]
Will review soon.

**Comment by instagibbs** [comment]
also publicly committing to reviewing soon

**Comment by jnewbery** [comment]
Manual test and very light code review ACK d26f0648f1c0d1115dcb8d76e57195032b88f400

I manually tested:

- default wallet is loaded on startup, even if "load_on_startup" is not set for the default wallet.
- if no default wallet exists, it's not created on startup
- this doesn't effect whether other wallets are loaded on startup or not.

**Review by jonatack — COMMENTED** [review]
ACK d26f0648f1c0d1115dcb8d76e57195032b88f400 light code review, debug build, ran tests, did manual testing with testnet, rebased on master, on linux debian.

One edge case seen: If I load a wallet (default or no), shutdown *without unloading it*, then `rm wallet.dat` or `rm -rf` the wallet folder, then restart, a new wallet is created in its place. 

Screenshot from 2020-09-17 14-51-15

**Comment by jonatack** [comment]
It seems possible to remove the additional "-wallet=" extra_arg from a number of the changed tests before finding a couple that fail without it.

Passing `-wallet` without a value also seems fine?
```diff
-            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet="],
-            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet="],
-            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet="],
-            ["-whitelist=noban@127.0.0.1", "-wallet="],
+            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet"],
+            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet"],
+            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet"],
+            ["-whitelist=noban@127.0.0.1", "-wallet"],
```

File: src/wallet/load.cpp
Comment (ryanofsky) [reply src/wallet/load.cpp:46]: re: https://github.com/bitcoin/bitcoin/pull/15454#discussion_r490224007

> Would it better to do:
> 
> ```diff
> -    if (!gArgs.IsArgSet("wallet")) {
> +    if (!(gArgs.GetBoolArg("wallet", false)) {
> ```

This would be bad because it would cause a `-nowallet` setting to load the default wallet.

**Review by ryanofsky — COMMENTED** [review]
re: https://github.com/bitcoin/bitcoin/pull/15454#issuecomment-694210398

> Passing `-wallet` without a value also seems fine?

Would not suggest this because that syntax more typically is used to set a boolean setting to true than to set a string setting to empty. #16545 would actually make it an error to use `-string` syntax for future string-only settings and would require `-string=` or `-string=""`

I think both versions of:

```diff
-            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet="],
+            ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet"],
```

are less clear than they could be, which is why the the followup https://github.com/ryanofsky/bitcoin/commit/5d2e85c5d33c3b65c79673cad5d7bc0e5875b107 I mentioned in 
 https://github.com/bitcoin/bitcoin/pull/15454#issuecomment-692747945 would change these to:

```c++
             ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet=" + self.default_wallet_name],
```

**Comment by jonatack** [comment]
> both versions...are less clear than they could be, which is why the the followup [ryanofsky@5d2e85c](https://github.com/ryanofsky/bitcoin/commit/5d2e85c5d33c3b65c79673cad5d7bc0e5875b107) I mentioned in
> [#15454 (comment)](https://github.com/bitcoin/bitcoin/pull/15454#issuecomment-692747945) would change these to:
> 
> ```c++
>              ["-whitelist=noban@127.0.0.1", "-keypool=100", "-wallet=" + self.default_wallet_name],
> ```

I agree, much better. Thanks @ryanofsky!

File: doc/release-notes-15454.md
Comment (instagibbs) [inline doc/release-notes-15454.md:4]: > Bitcoin Core will no longer create an unnamed `""` wallet by default.

Just stop here. It won't make a default wallet of any kind, whether or not you specify one specifically.

Comment (achow101) [reply doc/release-notes-15454.md:4]: I think you can make an unnamed wallet with `createwallet`

Comment (instagibbs) [reply doc/release-notes-15454.md:4]: > by default

is the key phrase here. It makes nothing by default.

Comment (achow101) [reply doc/release-notes-15454.md:4]: Meh. Release notes can always be edited when they go up on the dev wiki.

Comment (instagibbs) [reply doc/release-notes-15454.md:4]: Ok we'll just hope our readers are deficient in English ;P

Comment (flack) [reply doc/release-notes-15454.md:4]: maybe just leave out "by default"?

**Comment by meshcollider** [comment]
This is fine to go in as-is, as mentioned, we can easily make changes to the release notes when they are collated.

**Comment by Sjors** [comment]
I made some changes to the wallet creation dialog, because I don't think the current iteration is suitable for new users: https://github.com/bitcoin-core/gui/pull/96

**Comment by Sjors** [comment]
Forgot the extended tests... fixed in #19971

**Comment by ryanofsky** [comment]
re: https://github.com/bitcoin/bitcoin/pull/15454#issuecomment-694210398 & https://github.com/bitcoin/bitcoin/pull/15454#pullrequestreview-490569662

> It seems possible to remove the additional "-wallet=" extra_arg from a number of the changed tests

#20034 follows this PR up and removes all of these. It also goes further and removes the test framework `-wallet` extra_args mangling, which is no longer necessary thanks to this PR and #15937

**Pull request merged**
